woocommerce/woocommerce-android
# Hackweek/stats widget

Fixes the first four steps mentioned in #2041 for this hack week. 

**Note: The intention of this PR is to get everyone to play around with the changes and tests the widgets feature thoroughly. I wouldn't recommend making this PR live until we find a way to support multiple stores and multiple store notification support in the app.**

#### Changes
- Adds a new widget to the app that displays the current day stats, if v4 stats is available for the site. 
- Provides an option to users to select a site and the color mode of the widget.
- An error message is displayed in the widget if the user is not logged in to the app.
- An error message is displayed in the widget if the store does not have WC-Admin installed or if WooCommerce 4.0 is not installed.

#### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/84500417-1f82d900-acd2-11ea-8900-e3d31474bc2d.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500421-214c9c80-acd2-11ea-9461-d83b07bf1985.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500424-23aef680-acd2-11ea-9f67-6b97ca93c410.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500428-2578ba00-acd2-11ea-8b3c-347b43c37bbb.png" width="300"/>

<img src="https://user-images.githubusercontent.com/22608780/84500535-61138400-acd2-11ea-9c51-4e9e1480983d.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500541-640e7480-acd2-11ea-8c63-84e246df9836.png" width="300"/>
<img src="https://user-images.githubusercontent.com/22608780/84500586-7be5f880-acd2-11ea-8eb2-747a516f2fdc.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500592-7ee0e900-acd2-11ea-9b27-4348735fabbb.png" width="300"/>
<img src="https://user-images.githubusercontent.com/22608780/84500626-8dc79b80-acd2-11ea-9c9a-689d97f2f3d0.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500630-90c28c00-acd2-11ea-9886-8f2ba72506eb.png" width="300"/>
<img src="https://user-images.githubusercontent.com/22608780/84500654-9ddf7b00-acd2-11ea-8f7c-c75793edfc93.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/84500658-9fa93e80-acd2-11ea-95f5-dfb84b4d6cec.png" width="300"/>

#### Testing
- Run `TodayWidgetListViewModelTest` unit test.
- Add widget to a site that has WC-admin or Woo 4.0 available:
   - Login to a store that has WooCommerce 4.0 or WC-admin installed.
   - Long-press `home screen`.
   - Tap `Widgets` option.
   - Scroll to ***Woo*** widget.
   - Long press the ***Today*** widget and place on the home screen.
   - Note that the site selector screen is displayed.
   - Click on the site selector field and notice the list of sites displayed. Click on a site and verify that it is updated in the site selector field.
   - Click on the color selector field and notice a dialog is displayed with Light and Dark. Click on either and verify it is updated in the color selector field.
   - Notice the `Add Widget` button is not enabled.
   - Click on the button.
   - Notice the current day stats (visitors, order count and revenue) is displayed correctly in the widget.
- Verify that error message is displayed when user is not logged in.
   - Complete the previous test and add a widget to the app.
   - Open the app and logout of the app.
   - Close the app and notice that an error is displayed in the widget when you are logged out.
- Verify that error message is displayed when stats v4 is not available for a site:
   - Complete the first test and add a widget to the app.
   - Open the `wp-admin` of the current site you are logged into.
   - Disable WC-Admin plugin for the site and refresh the app.
   - Notice the stats reverted banner is displayed in the app.
   - Close the app and notice that the stats widget displays an error message.

#### Known issues
- I keep getting this [crash](https://gist.github.com/anitaa1990/b1e62dc14ea842209c033826bf25c007) when changing the system settings from dark mode to light mode or vice versa. I am yet to debug this crash.
- When the widget is clicked, we need to open the stats for the site assigned to the widget. The idea is to auto login the user to that site in the app. This is not yet implemented.
- Tracking events need to be added to this feature.

Update release notes:
